presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/util/TestRetryDriver.java

```diff
@@ -0,0 +1,127 @@
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import org.testng.annotations.Test;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+
+import static org.testng.Assert.fail;
+
+public class TestRetryDriver
+{
+    private static class FutureGenerator
+            implements Callable<ListenableFuture<Integer>>
+    {
+        private final int successAttempt;
+        private int currentAttempt;
+        private ListeningExecutorService executor;
+
+        public FutureGenerator(int successAttempt)
+        {
+            this.successAttempt = successAttempt;
+            this.currentAttempt = 0;
+            this.executor = MoreExecutors.listeningDecorator(Executors.newFixedThreadPool(1));
+        }
+
+        @Override
+        public ListenableFuture<Integer> call()
+                throws Exception
+        {
+            currentAttempt++;
+
+            if (currentAttempt < successAttempt && currentAttempt % 2 == 0) {
+                // Directly fail all even attempts before reaching the success attempt
+                throw new RuntimeException("Sync failure");
+            }
+
+            return executor.submit(() -> {
+                Thread.sleep(100);
+                if (currentAttempt >= successAttempt) {
+                    return 1;
+                }
+                else {
+                    throw new RuntimeException("Async failure");
+                }
+            });
+        }
+    }
+
+    @Test
+    void testOneAsyncRetryFail()
+            throws Exception
+    {
+        FutureGenerator futureGenerator = new FutureGenerator(2);
+        RetryDriver retryDriver = RetryDriver.retry().maxAttempts(1);
+
+        try {
+            Integer res = retryDriver.runAsync("test1", futureGenerator).get();
+            fail("Must fail due to no retry");
+        }
+        catch (ExecutionException ignore) {
+            // Must fail
+        }
+    }
+
+    @Test
+    void testOneAsyncRetry()
+            throws Exception
+    {
+        FutureGenerator futureGenerator = new FutureGenerator(2);
+        RetryDriver retryDriver = RetryDriver.retry().maxAttempts(2);
+
+        try {
+            Integer res = retryDriver.runAsync("test2", futureGenerator).get();
+        }
+        catch (ExecutionException ex) {
+            fail("Must succeed by the 2nd attempt");
+        }
+    }
+
+    @Test
+    void testMultipleAsyncRetryFail()
+            throws Exception
+    {
+        FutureGenerator futureGenerator = new FutureGenerator(4);
+        RetryDriver retryDriver = RetryDriver.retry().maxAttempts(3);
+
+        try {
+            Integer res = retryDriver.runAsync("test3", futureGenerator).get();
+            fail("Must failed after 3 attempts");
+        }
+        catch (ExecutionException ignore) {
+            // Must fail
```

Comment:
Do you test there actually were 3 attempts and that the returned exception makes sense (should it be failure of the first attempt, or last? or all (as suppressed exceptions)?)